stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -208,6 +214,97 @@ public void run() {
     });
   }
 
+  @ChromeDevtoolsMethod
+  public PerformSearchResponse performSearch(JsonRpcPeer peer, final JSONObject params) {
+    return mDOMProvider.postAndWait(new UncheckedCallable<PerformSearchResponse>() {
+      @Override
+      public PerformSearchResponse call() {
+        final PerformSearchRequest request = mObjectMapper.convertValue(
+            params,
+            PerformSearchRequest.class);
+
+        // Searches are case insensitive so convert
+        // the query to lower case for simpler comparisons.
+        final String lowerCaseQuery = request.query.toLowerCase();
+
+        final Object root = mShadowDOM.getRootElement();
+
+        // Each search action has a unique ID so that
+        // it can be queried later.
+        final String id = UUID.randomUUID().toString();
+
+        final ArrayList<Integer> results = new ArrayList<>();
+
+        findMatches(lowerCaseQuery, root, results);
+
+        mSearchResults.put(id, results);
+
+        final PerformSearchResponse response = new PerformSearchResponse();
+        response.searchId = id;
+        response.resultCount = results.size();
+
+        return response;
+      }
+    });
+  }
+
+  private void findMatches(String query, Object root, ArrayList<Integer> resultIds) {
+    final ElementInfo info = mShadowDOM.getElementInfo(root);
+
+    for (int i = 0, size = info.children.size(); i < size; i++) {
+      final Object element = info.children.get(i);
+
+      if (doesElementMatch(query, element)) {
+        resultIds.add(mObjectIdMapper.getIdForObject(element));
+      }
+
+      findMatches(query, element, resultIds);
+    }
+  }
+
+  private boolean doesElementMatch(final String query, final Object element) {
+      final NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(element);
+
+      final AttributeListAccumulator accumulator = getCachedAttributeListAccumulator();
+
+      try {
+        descriptor.getAttributes(element, accumulator);
+
+        for (int i = 0, size = accumulator.size(); i < size; i++) {
+          if (accumulator.get(i).toLowerCase().contains(query)) {
+            return true;
+          }
+        }
+      } finally {
+        returnCachedAttributeListAccumulator(accumulator);
+      }
+
+      return descriptor.getNodeName(element).toLowerCase().contains(query);
+  }
+
+  @ChromeDevtoolsMethod
+  public GetSearchResultsResponse getSearchResults(JsonRpcPeer peer, JSONObject params) {
+    final GetSearchResultsRequest request = mObjectMapper.convertValue(
+        params,
+        GetSearchResultsRequest.class);
+
+    final List<Integer> results = mSearchResults.get(request.searchId).subList(request.fromIndex, request.toIndex);
```

Comment:
We shouldn't trust that the `searchId` from the client is valid. If it's not then `mSearchResults.get()` will return null and we'll crash with an NPE. So let's make sure to check the return value from `get()` before using it. If it returns null then ... I guess we just return null? You can test that by just returning null from this method and making sure Chrome doesn't crash or do something weird (I doubt it will, but always good to double-check).